regex.c

```diff
@@ -248,58 +227,55 @@ void read_regex_from_file(void)
 	if(config.debug & DEBUG_REGEX)
 		regexbuffer = calloc(num_regex, sizeof(char*));
 
-	// Search through file
-	// getline reads a string from the specified file up to either a
-	// newline character or EOF
-	for(int i=0; getline(&buffer, &size, fp) != -1; i++)
+	// Connect to whitelist table
+	if(!gravityDB_getTable(REGEX_LIST))
+	{
+		logg("read_regex_from_database(): Error getting table from database");
+		return;
+	}
+
+	// Walk database table
+	const char *domain = NULL;
+	int i = 0;
+	while((domain = gravityDB_getDomain()) != NULL)
 	{
-		// Test if file has changed since we counted the lines therein (unlikely
-		// but not impossible). If so, read only as far as we have reserved memory
+		// Avoid buffer overflow if database table changed
+		// since we counted its entries
 		if(i >= num_regex)
 			break;
 
-		// Strip potential newline character at the end of line we just read
-		if(buffer[strlen(buffer)-1] == '\n')
-			buffer[strlen(buffer)-1] = '\0';
-
 		// Skip this entry if empty: an empty regex filter would match
 		// anything anywhere and hence match (and block) all incoming domains.
 		// A user can still achieve this with a filter such as ".*", however
 		// empty lines in regex.list are probably not expected to have such an
 		// effect and would immediately lead to "blocking the entire Internet"
-		if(strlen(buffer) < 1)
+		if(strlen(domain) < 1)
 		{
 			regexconfigured[i] = false;
-			logg("Skipping empty regex filter on line %i", i+1);
+			logg("Skipping empty regex filter");
 			skipped++;
 			continue;
 		}
 
-		// Skip this entry if it is commented out
-		if(buffer[0] == '#')
+		// Skip this entry if it is commented out using the legacy syntax
+		// (should use the disabled property of the table now)
+		if(domain[0] == '#')
 		{
 			regexconfigured[i] = false;
-			logg("Skipping commented out regex filter on line %i", i+1);
+			logg("Skipping commented out regex filter \"%s\"", domain);
 			skipped++;
 			continue;
 		}
 
-		// Compile this regex
-		regexconfigured[i] = init_regex(buffer, i);
-	}
-
-	// Free allocated memory
-	if(buffer != NULL)
-	{
-		free(buffer);
-		buffer = NULL;
+		// Copy this whitelisted domain into memory
+		regexconfigured[i] = init_regex(domain, i);
 	}
 
-	// Close the file
-	fclose(fp);
+	// Finalize statement and close gravity database handle
+	gravityDB_finalizeTable();
 
-	// Read whitelisted domains from file
-	read_whitelist_from_file();
+	// Read whitelisted domains from database
+	read_whitelist_from_database();
 
 	logg("Compiled %i Regex filters and %i whitelisted domains in %.1f msec (%i errors)", (num_regex-skipped), whitelist.count > 0 ? whitelist.count : 0, timer_elapsed_msec(REGEX_TIMER), errors);
```

Comment:
This method should only handle compiling regex. The regex-loading code (everything but the last two lines) should be moved into a separate function (and rename this function), such that the final function looks like this:
```cpp
read_regex_from_database();
read_whitelist_from_database();

logg(...);
```